### PR TITLE
evals: route iteration runners through prepareChatV2

### PR DESCRIPTION
**Pull request by chelojimenez**
## Summary

Eval ran its own tool-prep + system-prompt + temperature wiring, skipping `prepareChatV2`. That silently dropped skill tools, Anthropic name validation, skills-prompt assembly, and any progressive-discovery meta-tools from every eval run — users were evaluating a different surface than the one chat / playground / chatbox exposes.

This PR adopts `prepareChatV2` inside all four iteration runners — `runIterationWithAiSdk`, `runIterationViaBackend`, `streamIterationWithAiSdk`, `streamIterationViaBackend` — and threads `selectedServers` + `modelDefinition` through `RunIterationBaseParams` and the `runTestCase` / `streamTestCase` outer args. The engine (`generateText` / `streamText` / backend stream) and persistence (`persistEvalTraceFanout`, recorder, etc.) are unchanged.

This is **PR 1** of the engine-consolidation plan that lands eval on the same engine stack chat/playground/chatbox use. The plan lives at `~/.claude/plans/lets-consolidate-on-streamtext-transient-clarke.md`; subsequent PRs swap eval onto `runAssistantTurn` and collapse the four runners.

### Honest scoping on progressive discovery

`prepareChatV2` returns `{ allTools, progressivePlan, discoveryState }`. After this PR eval **advertises** the progressive-discovery meta-tools (`search_mcp_tools` / `load_mcp_tools`) when the plan is enabled — i.e. the meta-tools are now in the tool set the model sees. What this PR does **not** change:

- Eval's `prepareStep` still returns `undefined`. Chat's local path returns `activeTools` from `prepareStep` after a `gateToolsToActiveSubset` pass; eval skips that, so it advertises both the meta-tools and the full tool set every step instead of narrowing to the discovered subset.
- The backend runner serializes every entry from `prepared.allTools` into `toolDefs`, so the same lack-of-gating applies for hosted-model + hosted-org-BYOK paths.

Net effect: this PR closes the "skill tools / Anthropic validation / system prompt assembly / meta-tool presence" gap; it does **not** yet replicate chat's full progressive-exposure model. Active-tool gating lands when eval moves onto `runAssistantTurn` in a subsequent PR.

### Setup-failure behavior

`prepareChatV2` can throw on Anthropic name validation, meta-tool collisions, and skill-tool prep failures. Two adjustments to keep those visible:

- **AI-SDK runners** (`runIterationWithAiSdk` + `streamIterationWithAiSdk`): the `prepareChatV2` call is inside the existing outer try/catch, after iteration row creation. Setup throws land in the existing failed-iteration write path.
- **Backend runners** (`runIterationViaBackend` + `streamIterationViaBackend`): no outer try/catch existed. Added a small `persistSetupFailedIteration` helper; `prepareChatV2` is wrapped in a try/catch that records the failure (`status: "failed"`, `result: "failed"`, `error: <message>`) and returns early before the per-step fetch loop.

### What stays unchanged

- AI SDK call surface: `generateText` / `streamText` / Convex `/stream` + `/stream/org` backend loop
- Verdict + lock writer (`persistEvalTraceFanout`, `recorder.finalizeIteration`)
- `testIteration` row writes, configSnapshot, predicate verdicts
- Cancellation (AbortController + watcher unchanged)
- `wrapToolSetForEvalTrace` span capture
- `applyVisibilityPolicyAndCountSignals` at `runEvalSuiteWithAiSdk` level retained for `toolSignals` telemetry (`prepareChatV2` performs equivalent visibility filtering internally, but doesn't expose the dropped-count signals)

### Why `prepareChatV2` is one-shot per iteration (not per turn)

Chat-v2 calls `prepareChatV2` once per HTTP request — which is once per assistant turn, since each chat turn is its own request. Eval calls it once per iteration with `priorMessages: []`, meaning the progressive-discovery `discoveryState` only sees the initial empty state. A later PR (when eval moves to `runAssistantTurn`) naturally gets per-turn cadence because each turn becomes one driver call.

## Test plan

- [x] Typecheck: no new errors vs `main` baseline (4 pre-existing recorder shape issues, unrelated)
- [x] Eval-runner unit tests: 16 / 16 pass (15 originals + 1 new failure-path test asserting `prepareChatV2` throws are recorded as `status: "failed"` iteration rows)
- [x] All `server/services/evals` tests: 83 / 83 pass
- [x] Adjacent server suites (`utils/`, `services/`, `routes/web/`, `routes/mcp/`, `sessionSimulation/`): 759 / 759 pass — no chat/synthetic/route regressions
- [ ] End-to-end smoke on a local eval suite with a skill-enabled server — confirm skill tools appear in the transcript (today they don't)
- [ ] End-to-end smoke on a local eval suite with Anthropic + an MCP server exposing a tool with an invalid name — confirm the iteration is recorded as `status: "failed"` with the validation error, not as a suite-level crash

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes core eval execution paths and can alter which tools/prompts models see versus prior runs; failure handling fixes reduce silent pass inflation but behavior shifts are user-visible.
> 
> **Overview**
> Eval iteration runs now use the same **`prepareChatV2`** pipeline as chat for tools, system prompt, and temperature, instead of ad-hoc `getToolsForAiSdk` wiring. **`selectedServers`** and **`modelDefinition`** are threaded through **`runTestCase`**, **`streamTestCase`**, and all four runners (local AI SDK, backend, and their stream variants). Model calls use **`prepared.allTools`**, **`enhancedSystemPrompt`**, and **`resolvedTemperature`**.
> 
> **Setup failures** from `prepareChatV2` (e.g. Anthropic tool-name validation) are handled explicitly: backend paths get **`persistSetupFailedIteration`** and early return; streaming backend paths also emit an SSE **`error`** event. Returned **`evaluation.passed`** is forced to **`false`** on setup and outer-catch failures so suite pass counts are not inflated (including negative tests with no tool calls).
> 
> Suite-level **`tools`** remains only for **`toolSignals`** telemetry. Tests mock **`prepareChatV2`** and add coverage for failed setup persistence, negative-test pass counting, and stream error emission.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4928abde6a857d724ce910a21fd7d814563fde16. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/86bcb804-db4f-4579-b71a-bdc4f1c26242?refs=4928abde6a857d724ce910a21fd7d814563fde16&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: mcpjam-inspector/server/services/evals-runner.ts [hunk 1]
```diff
@@ -1215,7 +1246,7 @@ const runIterationWithAiSdk = async ({
       activeCompletedStepCount = 0;
       activeTraceCtx = createAiSdkEvalTraceContext(runStartedAt);
       const tracedTools = wrapToolSetForEvalTrace(
-        tools,
+        prepared.allTools,
         activeTraceCtx,
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/server/services/evals-runner.ts:1292]: Acknowledged — this is the same correctness gap Marcelo flagged in the human review, and the PR body's "Honest scoping on progressive discovery" section captures it explicitly:

> Eval's `prepareStep` still returns `undefined`. Chat's local path returns `activeTools` from `prepareStep` after a `gateToolsToActiveSubset` pass; eval skips that, so it advertises both the meta-tools and the full tool set every step instead of narrowing to the discovered subset. The backend runner serializes every entry from `prepared.allTools` into `toolDefs`, so the same lack-of-gating applies for hosted-model + hosted-org-BYOK paths.

Net effect for this PR: meta-tools are advertised, but active-tool gating is not implemented. Closing the gating gap lands when eval moves onto `runAssistantTurn` in a follow-up PR (chat's `gateToolsToActiveSubset` + `prepareStep` machinery comes for free with that swap).

Not changing this PR's behavior here.

**Comment by chelojimenez** [comment]
## Bot review triage (commit `622e2a8`)

| Comment | Verdict | Action |
|---|---|---|
| **Cursor — backend `customProviders` omission** (Medium) | Valid. Anthropic name validation was skipped for hosted-org BYOK paths using Anthropic-compatible custom providers. | **Fixed.** Backend runners now derive `customProviders` from `orgModelConfig` via `buildLlmRuntimeConfigFromOrgConfig` and pass to `prepareChatV2`. Matches the local-AI-SDK runner. |
| **CodeRabbit — same asymmetry** (nitpick) | Same issue. | **Subsumed by above fix.** |
| **Cursor — setup failure reports passing evaluation** (Medium) | Valid. `evaluateMultiTurnResults([], ...)` returns `passed: true` for negative tests and for positive tests with no expected tools; `runEvalSuiteWithAiSdk` aggregates `evaluation.passed`, so setup failures silently counted as suite passes. | **Fixed.** Both backend setup-failure paths and both AI-SDK runners' outer catch blocks force `evaluation.passed = false` before returning. Added regression test for the negative-test edge case. |
| **Codex — progressive-discovery filtering not applied** (P2) | Pre-existing scope decision. Same correctness gap Marcelo flagged in the human review. Explicitly deferred. | **No code change.** Active-tool gating (chat's `gateToolsToActiveSubset` + `prepareStep` machinery) lands when eval moves onto `runAssistantTurn` in a follow-up PR. PR body's "Honest scoping on progressive discovery" section already captures this. |

**Tests:** 17 / 17 eval-runner tests pass (15 originals + the previous setup-failure test + the new negative-test regression guard). 760 / 760 server-side tests across `services/`, `utils/`, `routes/web/`, `routes/mcp/` pass.

**Typecheck:** no new errors vs `main` baseline.

File: mcpjam-inspector/server/services/evals-runner.ts [hunk 2]
```diff
@@ -1659,7 +1738,48 @@ const runIterationViaBackend = async ({
...
+      recorder,
+      convexClient,
+    });
+    return {
+      evaluation: evaluateMultiTurnResults(
+        promptTurns,
+        [],
+        test.isNegativeTest,
+        test.matchOptions,
+      ),
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/server/services/evals-runner.ts]: This was reviewed on the older commit `57619c7` and is a duplicate of Cursor Bugbot's "Setup failure reports passing evaluation" — same root cause (`evaluateMultiTurnResults([], …)` returning `passed: true` on negative tests / zero-expected-tool tests). Already fixed in commit `622e2a8` with regression test:

- Both backend setup-failure paths and both AI-SDK runners' outer catch blocks force `evaluation.passed = false` before returning.
- `does not count a negative-test setup failure as a suite pass` test asserts a negative test with `expectedToolCalls: []` lands as `status:"failed"` / `result:"failed"` even when `evaluateMultiTurnResults` would otherwise have returned `passed: true`.

No additional code change needed.

File: mcpjam-inspector/server/services/evals/__tests__/evals-runner.test.ts [hunk 3]
```diff
@@ -943,4 +965,77 @@ describe("runEvalSuiteWithAiSdk compare session metadata", () => {
...
+    expect(payload.result).toBe("failed");
+    expect(payload.error).toEqual(expect.stringContaining("Invalid tool name"));
+    expect(payload.iterationId).toBe("iter-failed-setup");
+
+    // The runner must NOT have called generateText — the failure happens
+    // before any model invocation.
+    expect(generateTextMock).not.toHaveBeenCalled();
+
+    prepareSpy.mockRestore();
+  });
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/server/services/evals/__tests__/evals-runner.test.ts:1044]: Good catch — applied in commit `3924d0c`. Wrapped both setup-failure tests in `try/finally` so `prepareSpy.mockRestore()` always runs. Verified 17/17 tests still pass.

**Comment by chelojimenez** [comment]
## Bot review triage round 2 (commit `3924d0c`)

| Comment | Verdict | Action |
|---|---|---|
| **Codex P2 — "Force setup failures to count as failed"** (on stale commit `57619c7`) | Duplicate of Cursor's "Setup failure reports passing evaluation" from the same review round. Already fixed in `622e2a8` with regression test for the negative-test edge case. | **No code change.** Replied in thread. |
| **CodeRabbit — `prepareSpy.mockRestore()` not exception-safe** (Minor / Quick win, on `622e2a8`) | Valid. If any assertion threw before the trailing `mockRestore()`, the spy would leak into later tests. Same pattern in both my setup-failure tests. | **Fixed.** Both tests wrapped in `try/finally` so restore always runs. 17/17 tests still pass. |

File: mcpjam-inspector/server/services/evals-runner.ts [hunk 4]
```diff
@@ -3310,7 +3502,65 @@ const streamIterationViaBackend = async ({
...
+      promptTurns,
+      [],
+      test.isNegativeTest,
+      test.matchOptions,
+    );
+    failedEvaluation.passed = false;
+    return {
+      evaluation: failedEvaluation,
+      iterationId,
+    };
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/server/services/evals-runner.ts:3569]: Good catch — fixed in commit `4928abd`. Added `emit({type:"error", message: errorMessage})` to the `streamIterationViaBackend` setup-failure catch, right after `persistSetupFailedIteration`. Now matches `streamIterationWithAiSdk`'s in-stream failure signal for hosted-model / hosted-org-BYOK runs.

Regression test asserts (a) the `error` event fires on the SSE stream, (b) the iteration row is still persisted as `status:"failed"`, and (c) the per-step fetch is never called. 18/18 eval-runner tests pass, 761/761 server-side tests pass.
